### PR TITLE
Headers Override for Name and CrowdAnkiUUID

### PR DESCRIPTION
**Pull request by ohare93**
Added the ability to override name and crowdanki_uuid for a Deck Header.

For https://github.com/anki-geo/ultimate-geography/pull/566

### CODE REVIEW

**Comment by aplaice** [comment]
Thanks so much for the fast implementation! It works! :)

Isn't a (slight) design issue that this prevents BrainBrew from being able to (even in principle) import changes to the deck name (and the `crowdanki_uuid`)? (when running anki_to_source)

(BrainBrew currently doesn't allow writing to the YAML headers file, but in principle it's possible, while if the name and crowdanki_uuid are hard-coded in the recipes then it becomes impossible even in principle.)

(I might be overthinking this, though — `anki_to_source` is currently probably only rarely used, and changing the deck name (and especially the deck UUID) from the Anki side will be _extremely_ rare.)

File: brain_brew/build_tasks/overrides/headers_override.py [hunk 1]
```diff
@@ -17,25 +17,39 @@ def task_name(cls) -> str:
...
             deck_description_html_file: str(required=False)
+            crowdanki_uuid: str(required=False)
+            name: str(required=False)
         '''
 
     @dataclass
     class Representation(RepresentationBase):
         deck_description_html_file: Optional[str]
+        crowdanki_uuid: Optional[str]
+        name: Optional[str]
```
Comment (aplaice) [inline brain_brew/build_tasks/overrides/headers_override.py:28]: I think that we need default values (`None`) here, if we want to allow overriding only some of `deck_description`, `crowdanki_uuid` and `name` at the same time.

```suggestion
        deck_description_html_file: Optional[str] = None
        crowdanki_uuid: Optional[str] = None
        name: Optional[str] = None
```

Comment (ohare93) [reply brain_brew/build_tasks/overrides/headers_override.py:28]: I don't believe so, since the actual overriding function individually check if each value is truthy. Have you experienced that populating one in the recipe, and not others, will null out the others?

Comment (aplaice) [reply brain_brew/build_tasks/overrides/headers_override.py:28]: Overriding one in the recipe, but not the others causes BrainBrew to crash:

```
TypeError: __init__() missing 2 required positional arguments: 'crowdanki_uuid' and 'name'
```

**Comment by ohare93** [comment]
> Thanks so much for the fast implementation! It works! :)

No worries, any time :grin: 

> Isn't a (slight) design issue that this prevents BrainBrew from being able to (even in principle) import changes to the deck name (and the `crowdanki_uuid`)? (when running anki_to_source)

Not sure I understand what you are saying. That there should be a way to programmatically write to a Header?

> (BrainBrew currently doesn't allow writing to the YAML headers file, but in principle it's possible, while if the name and crowdanki_uuid are hard-coded in the recipes then it becomes impossible even in principle.)

Doesn't `headers_from_crowd_anki` do this? That is what is used in the `brain_brew init` function to setup a repo. Or am I misunderstanding? 😅

File: brain_brew/build_tasks/overrides/headers_override.py (hunk 1 above)
Comment (ohare93) [reply brain_brew/build_tasks/overrides/headers_override.py]: Huh, I guess you're right. I never made that default value in the original since it was the only member of the Override class. Pushed a change that should fix it :+1:

**Comment by aplaice** [comment]
I think that I'm overthinking this (firstly, changing headers is a very rare use-case to begin with, and secondly changing the name and UUID makes little sense and/or will break other things) so please feel free to ignore the below! (I'm not quite sure why I decided to type all of this up, given that I realised half-way through that it makes no sense...)

<hr/>

<details><summary>Overlong argument</summary>

> Doesn't headers_from_crowd_anki do this? That is what is used in the brain_brew init function to setup a repo. Or am I misunderstanding? sweat_smile

I meant that currently (at least in AUG's `anki_to_source.yaml`) changes to the deck header fields (`extendNew`, `extendRev`, `name`) in `build/Ultimate Geography [EN]/deck.json` don't propagate to `src/headers/default.yaml` when running `recipes/anki_to_source.yaml`.

Now actually looking at `anki_to_source.yaml` that's because `save_to_file: src/headers/default.yaml` is commented out... (So it's possible both in principle and in practice, but (rightly) commented out because it's rarely used.)

> Not sure I understand what you are saying. That there should be a way to programmatically write to a Header?

I meant that now that `name` and `crowdanki_uuid` are (at least in some cases) hardcoded in the recipes, they can't (in these cases) be updated when using `headers_from_crowdanki`, even in principle, in a way that the deck description still could, in principle. (`deck_description_html_file` is a "pointer", while `name` and `crowdanki_uuid` store the raw values.)

### `deck_description_html_file` current behaviour

Currently, when `deck_description_html_file` is "overriden", then the deck description also can't be programmatically changed. If you change `desc` in a `deck.json`, run `headers_from_crowd_anki` (e.g. like in `recipes/anki_to_source.yaml`, when you uncomment `save_to_file: src/headers/default.yaml`), then you end up with `desc: changed description` in `src/headers/default.yaml` (I'm using the paths as they are in AUG), but given that you still have

```
        override:
          deck_description_html_file: src/headers/desc.html
```

in `recipes/source_to_anki.yaml`, the changed description (in `src/headers/default.yaml`) won't ever be used. (So if you change `desc`, run `recipes/anki_to_source.yaml` (with `save_to_file:` uncommented) and run `recipes/source_to_anki`, you'll end up with your original `desc` in `deck.json`. (Obviously, this is without manually modifying in the meantime!))

### `deck_description_html_file` theoretically possible behaviour

(Note: this is not a request for this feature, just for illustration!)

However, one can imagine having an `override` field in `headers_from_crowd_anki` (I don't think it's currently possible), for example something like:

```
  - headers_from_crowd_anki:
    - source: build/Ultimate Geography [EN]
      part_id: default
      save_to_file: src/headers/default.yaml
      override:
          deck_description_html_file: src/headers/desc.html
```

such that, when you run this `headers_from_crowd_anki` (as part of a `anki_to_source.yaml` recipe), the `desc` field from `deck.json` is written to `src/headers/desc.html`. (Consequently, if you change `desc` in a `deck.json` and run `recipes/anki_to_source.yaml` and then `recipes/source_to_anki.yaml` you still have your changed `desc`.)

### `crowdanki_uuid` and `name`

In contrast, given that the values of `crowdanki_uuid` and `name` are being stored in the recipe itself, if you were to change the `name` or `crowdanki_uuid` in `deck.json`, there is no way for them to be programmatically updated (in the way that we could, in principle, update the contents of the `src/headers/desc.html` file).

However, on second thought this doesn't matter.

If the `name` is changed, then the path to the build directory will have changed (since it's `build/${name}/deck.json`), so the anki_to_source recipe won't work anyway (without manual editing).

`crowdanki_uuid` should not ever change, since it's the id identifying the deck.

</details>

<hr/>

tldr; everything is fine, ignore the above.

**Comment by aplaice** [comment]
On a more constructive note, the partial overrides issue is now fixed (there's no crash when only some fields are overridden)!

**Comment by aplaice** [comment]
AFAICT this works perfectly (as we'd require for #566) with all even minor issues (key sorting) dealt with!

**Comment by aplaice** [comment]
@ohare93 could you please merge this? (Not urgently!)

(I haven't tested it any more, but from what I remember from https://github.com/anki-geo/ultimate-geography/pull/566 it worked exactly as needed.)

**Comment by ohare93** [comment]
Huh I never merged this? Whoops! Done 😁 I'll release a new version shortly 👍

**Comment by ohare93** [comment]
Released 👍 https://github.com/ohare93/brain-brew/releases/tag/v0.3.10

**Comment by ohare93** [comment]
@omarkohl I see you referenced this PR. It is released now, apologies for the delay! 😅

**Comment by aplaice** [comment]
Thanks!
